# Pre-lecture HW

## Q1

In [6]:
import pandas as pd
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
df = pd.read_csv(url)
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Q2

After Trying to acomplishing task 1 in prelecture HW, Chatgpt gave me a code to run as anwser to my question since it couldn't open the dataset himeself

here is the code:

In [7]:
import pandas as pd

# Load the Titanic dataset from your local file
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Get the columns of the dataset and the size (rows, columns)
columns = titanic_data.columns
size = titanic_data.shape

# Display the columns and the size
print(f"Columns: {columns}")
print(f"Size: {size}")

Columns: Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
Size: (891, 12)


# Q 2.2

Well to be fair GPT gave me a really well rounded anwser but I think i deserve some creddit for good prompts =)

in the context of my dataset: 

- Observations = rows = individual passengers.
- Variables = columns = characteristics or data points about each passenger.

I discussed the output with ChatGPT, since I didn't exactly know what "Size: (891, 12)" stood for, the data set -even though with some missing values- is a dataset of 891 Titanic passengers and 12 characteristics of each individual represented in each column

## Q3

After someback and forth between me and GPT we landed on a decent qway to sumrize the ccolumns

1. For numerical columns, I'll provide basic statistics like mean, min, max, and missing values.
2. For categorical columns (like Sex, Embarked), I'll show the distribution of unique values and missing data.

note that in the list above I refers to ChatGPT

One neat problem was the Survived column which was technicly numeric but it used 0s and 1s as boolians

simllarly to one before, Pclass is only 1s 2s and 3s, again technicly numeric but it refers to Classes not actual numbers

I asked GPT to adjust for thoose 

Here is the code:

In [8]:
import pandas as pd

# Load the Titanic dataset from your local file
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Basic summary for each column including missing values
def summarize_column(col):
    print(f"\nSummary for column: {col.name}")
    print(f"  Data type: {col.dtype}")
    print(f"  Missing values: {col.isnull().sum()}")

    # Special handling for Survived (binary column)
    if col.name == 'Survived':
        print(f"  Survival distribution:\n{col.value_counts()}")
        survival_rate = col.mean() * 100
        print(f"  Survival rate: {survival_rate:.2f}%")
    
    # Special handling for Pclass (categorical class column)
    elif col.name == 'Pclass':
        print(f"  Passenger Class distribution:\n{col.value_counts()}")
    
    # If numeric, summarize like normal numeric data
    elif pd.api.types.is_numeric_dtype(col):
        print(f"  Mean: {col.mean():.2f}")
        print(f"  Min: {col.min()}")
        print(f"  Max: {col.max()}")
        print(f"  Standard Deviation: {col.std():.2f}")
    else:
        # Otherwise, treat it as categorical and show unique values
        print(f"  Unique values:\n{col.value_counts(dropna=False)}")

# Loop through all columns in the dataset and summarize
for col_name in titanic_data.columns:
    summarize_column(titanic_data[col_name])


Summary for column: PassengerId
  Data type: int64
  Missing values: 0
  Mean: 446.00
  Min: 1
  Max: 891
  Standard Deviation: 257.35

Summary for column: Survived
  Data type: int64
  Missing values: 0
  Survival distribution:
Survived
0    549
1    342
Name: count, dtype: int64
  Survival rate: 38.38%

Summary for column: Pclass
  Data type: int64
  Missing values: 0
  Passenger Class distribution:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Summary for column: Name
  Data type: object
  Missing values: 0
  Unique values:
Name
Braund, Mr. Owen Harris                     1
Boulos, Mr. Hanna                           1
Frolicher-Stehli, Mr. Maxmillian            1
Gilinski, Mr. Eliezer                       1
Murdlin, Mr. Joseph                         1
                                           ..
Kelly, Miss. Anna Katherine "Annie Kate"    1
McCoy, Mr. Bernard                          1
Johnson, Mr. William Cahoone Jr             1
Keane, Miss. Nora A             

## Q4

Once again, Hats off to GPT for giving such complite anwser im basicly just writing it in md, but here it is:

- **Non-numeric variables are excluded from df.describe()**
    - If you have 12 columns but 5 of them are non-numeric, df.describe() will only analyze the remaining 7 numeric columns, leading to fewer columns in the describe() output.
    - df.describe(), by default, only analyzes numeric variables, so it will exclude non-numeric (categorical) columns like Name, Sex, Ticket, Cabin


- **Missing numeric values will not be analyized by df.describe()**
    - df.describe() gives summaries like mean, min, max, and counts only non-missing (valid) entries for each numeric column.
    - in casse of our dataset, the analisis above is only for existing ages since our set has a lot of missing ages, the min max and mean are for only the valid ones
    
- df.shape reports the total number of rows and columns in your dataset.
    
to oversimlify it but in my own words, in case of non-numeric vars, df.describe() will not analyze and in case of missing numeric vars df.describe() will only count the values that are there. To the contrary df.shape will count every row and column regardless of it being numeric or not and missing values or not it will give the overall number of rows and columns

## Q5

#### Method:
- A method is a **function** that is associated with an object, and it can perform operations or actions on the object’s data, typically modifying or providing a new view of the data.

#### Attribute:
- An attribute is a **property** of an object that stores data or state information about the object.

above you see the gpt responce but it think it's best said that an atribiute like shape is something that in oop is associathed with our object at initilazation and the maybe changed but a method is a functtion that we can use on our object.

methods might take arguments and perform an operation but attribut are accessed directly from the object itself.

# Chatbot summary pt.1 (1-5)

Here is the link to my ChatGPT intraction :
https://chatgpt.com/share/dcb58cec-d4bd-4c51-8b26-59c131f95bad

[Click Here](https://chatgpt.com/share/dcb58cec-d4bd-4c51-8b26-59c131f95bad)

---

#### Understanding the Titanic Dataset Structure

Objective: To gain a foundational understanding of the Titanic dataset's structure and how to interpret its dimensions and variables.

#### Dataset Dimensions:

- Rows (Observations): Each of the 891 rows represents an individual passenger aboard the Titanic.
- Columns (Variables): The 12 columns represent different characteristics (variables) describing each passenger (e.g., PassengerId, Survived, Pclass, Age).

#### Explaining Observations and Variables:

- Observations: Refers to individual entries in the dataset; in this case, individual Titanic passengers.
- Variables: Refers to the features or attributes recorded for each observation, such as survival status, class, and age.

#### Discrepancies in Data Analysis:

- Non-Numeric Variables: df.shape reports all columns, but df.describe() by default only includes numeric variables. This results in fewer columns being analyzed by df.describe().
- Missing Values in Numeric Variables: df.shape counts all rows, but df.describe() counts only non-missing entries for each column. This can lead to discrepancies in the "count" between these two methods, particularly noticeable in columns with missing data like Age.

#### Summary Statistics:

Discussed how to generate and interpret summary statistics for both numeric and categorical data within the dataset, adjusting the approach based on the nature of each variable (e.g., treating Survived as a categorical variable).

#### Code Adjustments:
Adjusted Python code to correctly summarize the dataset, accounting for the specific nature of binary and categorical columns like Survived and Pclass.

#### Conclusion:
The Titanic dataset was dissected to understand its structure and the characteristics of its variables. Practical steps to summarize the dataset were discussed, and Python code was provided and adjusted to handle specific data types accurately.

# Post lecture HW

- note: since i did the diffrent sections on diffrent days i had 2 diffrent GPT settions but i tried my best to give the context of my first settion to my second chat

## Q6

fairly straightforward

1. count: The number of non-missing (non-NaN) entries for each variable.
2. mean: The average value of the variable.
3. std: The standard deviation, a measure of the dispersion or spread of the values.
4. min: The minimum value of the variable.
5. 25%: The first quartile (Q1), which is the value below which 25% of the data falls.
6. 50%: The median (second quartile, Q2), which is the middle value when data is sorted.
7. 75%: The third quartile (Q3), which is the value below which 75% of the data falls.
8. max: The maximum value of the variable.

## Q7

1. Dataset with multiple columns, and we want to remove rows that contain missing values in any of these columns because those rows would lead to inaccurate analysis.This method is preferred because it preserves the structure of all columns but removes only those rows where one or more cells are missing data.

2. Large dataset with some columns containing many missing values, and those columns are deemed unnecessary or irrelevantfor analysis.This method is preferred when we decide to drop specific columns that are either not needed or have too many missing values.

3. Applying del df['col'] before df.dropna() can be crucial because removing irrelevant or highly missing columns first can prevent the unnecessary loss of valuable data in other columns.

4. approach:

- Remove the Cabin Column Entirely: Since the Cabin column has a significant number of missing values and is generally less informative for most analyses (e.g., predicting survival), it makes sense to delete this column altogether. 

- Handle the Age Column Separately


I asked GPT to write the code for this aproach and provide a before and after summary =)

In [9]:
import pandas as pd

# Load the Titanic dataset from the provided URL
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Function to provide a simplified report on the dataset
def simplified_report(df, title):
    print(f"\n{title} Report:")
    print(f"  Total Rows: {df.shape[0]}")
    print(f"  Total Columns: {df.shape[1]}")
    print(f"  Missing Values Per Column:\n{df.isnull().sum()}")

# Initial simplified report before data cleaning
simplified_report(titanic_data, "Before Cleaning")

# Step 1: Delete the 'Cabin' column due to a high number of missing values and low relevance
del titanic_data['Cabin']

# Step 2: Impute missing values in 'Age' with the median value
titanic_data['Age'].fillna(titanic_data['Age'].median(), inplace=True)

# Step 3: Drop rows with any remaining missing values
titanic_data.dropna(inplace=True)

# Simplified report after data cleaning
simplified_report(titanic_data, "After Cleaning")



Before Cleaning Report:
  Total Rows: 891
  Total Columns: 12
  Missing Values Per Column:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

After Cleaning Report:
  Total Rows: 889
  Total Columns: 11
  Missing Values Per Column:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


## Q8 - 1

In my own Words:

what "df.groupby("col1")["col2"].describe()" does is generate and show "count, mean, ..., max" of col2, for each unique value of col1. 

#### Summary of Imaginary Dataset and Demonstration

- We created an imaginary dataset to analyze student GPAs from four different universities: University of Toronto (UofT), University of Waterloo (UWaterloo), York University (York), and Western University (Western). Each university had 20 students, making a total of 80 students in the dataset.

- Dataset Structure:
    - University: Represents the name of the university (UofT, UWaterloo, York, Western).
    - GPA: Represents the GPA of each student, randomly generated based on a normal distribution with different means and standard deviations for each university.

- Objective:
    - To demonstrate the use of df.groupby("col1")["col2"].describe() to analyze the distribution of GPAs (GPA) for students from each university (University).

Here is the ChatGPT code bellow:

In [10]:
import pandas as pd
import numpy as np

# Set a random seed for reproducibility
np.random.seed(42)

# Create an imaginary dataset with 100 students (20 from each university)
data = {
    'University': (['UofT'] * 20) + (['UWaterloo'] * 20) + (['York'] * 20) + (['Western'] * 20),
    'GPA': np.concatenate([
        np.random.normal(3.5, 0.3, 20),  # GPAs for UofT students
        np.random.normal(3.2, 0.4, 20),  # GPAs for UWaterloo students
        np.random.normal(2.8, 0.5, 20),  # GPAs for York students
        np.random.normal(3.0, 0.6, 20)   # GPAs for Western students
    ])
}

# Create a DataFrame
df_students = pd.DataFrame(data)

# Group by 'University' and describe the 'GPA' column
gpa_description_by_university = df_students.groupby("University")["GPA"].describe()

# Display the result
print(gpa_description_by_university)


            count      mean       std       min       25%       50%       75%  \
University                                                                      
UWaterloo    20.0  3.093610  0.387215  2.416132  2.767587  3.096506  3.279945   
UofT         20.0  3.448610  0.288009  2.926016  3.305384  3.429756  3.652453   
Western      20.0  2.981162  0.667253  1.428153  2.687604  3.015366  3.488922   
York         20.0  2.786655  0.410424  1.918480  2.542645  2.813930  3.121687   

                 max  
University            
UWaterloo   3.940911  
UofT        3.973764  
Western     3.938786  
York        3.328561  


## Q8 - 2

#### Difference Between `df.describe()` and `df.groupby("col1")["col2"].describe()` Counts:

1. **`df.describe()` Count**:
   - Shows the number of non-missing (non-NaN) values **for each column** in the entire dataset.
   - If a column has missing data, its `count` will be less than the total number of rows. This tells you how much data is available for each column overall.

2. **`df.groupby("col1")["col2"].describe()` Count**:
   - Shows the number of non-missing values **for `col2` within each group** defined by the unique values of `col1`.
   - It tells you how much data is available for `col2` in each category of `col1` (e.g., grouping by passenger class and counting ages in the Titanic dataset).

#### Key Difference:
- `df.describe()` provides a summary of missing data **for each column** in the entire dataset.
- `df.groupby("col1")["col2"].describe()` provides a summary of missing data **for `col2` within each group** of `col1`.

#### Why This Matters:
- `df.describe()` helps you see overall data completeness for each column.
- `df.groupby("col1")["col2"].describe()` helps you understand data completeness and distribution **within specific groups**, which is useful for more detailed analysis.

This distinction is important because one summarizes data for the whole dataset, while the other focuses on grouped subsets.


# Q8 -3

## A

In [11]:
# Omitted: import pandas as pd

# Create a DataFrame without importing pandas
data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
df = pd.DataFrame(data)  # This will raise an error

In [12]:
import pandas as pd

# Now this works correctly
data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
df = pd.DataFrame(data)
print(df)

   A  B
0  1  4
1  2  5
2  3  6


## B

In [13]:
import pandas as pd

# Mistyped filename in the URL
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanics.csv'  # Incorrect filename 'titanics.csv'
titanic_data = pd.read_csv(url)  # This will raise an error


HTTPError: HTTP Error 404: Not Found

In [ ]:
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'  # Correct filename 'titanic.csv'
titanic_data = pd.read_csv(url)

## C

In [ ]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

In [14]:
# Incorrectly using the DataFrame variable
# Intentional typo in the variable name (should be 'titanic_data')
Titanic_Data.groupby("Pclass")["Age"].describe()  # This will raise an error

NameError: name 'Titanic_Data' is not defined

In [15]:
# Correctly using the DataFrame variable
titanic_data.groupby("Pclass")["Age"].describe()  # Correct usage

count       mean        std   min    25%   50%   75%   max
Pclass                                                            
1       214.0  36.688879  14.142850  0.92  28.00  35.0  47.0  80.0
2       184.0  29.765380  13.581096  0.67  23.00  28.0  36.0  70.0
3       491.0  25.932627  10.697676  0.42  20.25  28.0  29.0  74.0

## D

In [16]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# Forgetting the closing parenthesis in the read_csv function
titanic_data = pd.read_csv(url  # Missing closing parenthesis


SyntaxError: incomplete input (1838234429.py, line 7)

In [17]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# Correcting by adding the closing parenthesis
titanic_data = pd.read_csv(url)  # Now this works correctly

## E

In [18]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Mistyping 'groupby' as 'group_by'
titanic_data.group_by("Pclass")["Age"].describe()  # This will raise an error


AttributeError: 'DataFrame' object has no attribute 'group_by'

In [19]:
# Correcting the function name 'groupby'
titanic_data.groupby("Pclass")["Age"].describe()  # Now this works correctly

count       mean        std   min   25%   50%   75%   max
Pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [20]:
# Mistyping 'describe' as 'describle'
titanic_data.groupby("Pclass")["Age"].describle()  # This will raise an error

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

In [21]:
# Correcting the function name 'describe'
titanic_data.groupby("Pclass")["Age"].describe()  # Now this works correctly

count       mean        std   min   25%   50%   75%   max
Pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

## F

In [34]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Mistyping the column name 'Sex' as 'sex' in groupby
titanic_data.groupby("sex")["Age"].describe()  # This will raise an error

KeyError: 'sex'

In [33]:
# Correct column names for groupby and selection
titanic_data.groupby("Sex")["Age"].describe()  # Correct groupby column 'Sex'

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [24]:
# Mistyping the column name 'age' as 'Age' in the column selection
titanic_data.groupby("sex")["age"].describe()  # This will raise an error

KeyError: 'sex'

In [32]:
# Correct column names for groupby and selection
titanic_data.groupby("Sex")["Age"].describe()  # Correct column selection 'Age'

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

## G

In [26]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Missing quotes around 'sex' in groupby
titanic_data.groupby(sex)["Age"].describe()  # This will raise an error


NameError: name 'sex' is not defined

In [30]:
# Correct column name usage with quotes
titanic_data.groupby("Sex")["Age"].describe()  # Correct usage with quotes

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [28]:
# Missing quotes around 'Age' in column selection
titanic_data.groupby("sex")[Age].describe()  # This will raise an error

KeyError: 'sex'

In [31]:
# Correct column name usage with quotes
titanic_data.groupby("Sex")["Age"].describe()  # Correct usage with quotes

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

## Troubleshooting Code: ChatGPT vs. Google

Alright, here’s the deal. When I’m stuck with coding issues, I’d rather use ChatGPT over Google. Why? Because with Google, I have to dig through a bunch of websites, read endless threads, and sometimes still not find exactly what I need. It’s like a scavenger hunt.

With ChatGPT, it’s more like having a friend who knows their stuff and walks me through the problem step-by-step. It’s way more direct and less frustrating. Plus, I can ask follow-up questions without jumping from page to page. So yeah, ChatGPT feels more like a human tutor, making troubleshooting less of a headache.

## Q9

**Somewhat**

# Chatbot summary pt.2 (6-9)

Here is the link to my ChatGPT intraction :
https://chatgpt.com/share/5119ba84-18b7-4a42-a1a3-27335f143ebb

[Click Here](https://chatgpt.com/share/5119ba84-18b7-4a42-a1a3-27335f143ebbd)

### Chatbot Summary: Data Cleaning and Error Handling with the Titanic Dataset

#### Objective:
To clean the Titanic dataset by handling missing values effectively and then intentionally introduce and fix common coding errors to improve data analysis and debugging skills.

#### Key Steps Taken:

1. **Understanding the Data**:
   - Reviewed the Titanic dataset's structure, identifying that the `Cabin` and `Age` columns had a significant number of missing values, while most other columns had very few or none.

2. **Decided on the Data Cleaning Approach**:
   - **Remove the `Cabin` column**: Due to a high number of missing values (687 out of 891) and its low relevance to most analyses.
   - **Impute missing values in the `Age` column**: Fill missing `Age` values with the median to maintain data integrity without introducing bias.
   - **Drop remaining rows with missing values**: After handling `Cabin` and `Age`, drop any rows with missing values in other columns to ensure a clean dataset.

3. **Implemented the Data Cleaning Steps**:
   - Removed the `Cabin` column using `del df['Cabin']`.
   - Imputed missing `Age` values using the median with `df['Age'].fillna(df['Age'].median(), inplace=True)`.
   - Dropped any remaining rows with missing values using `df.dropna(inplace=True)`.

4. **Created a Simplified Before and After Report**:
   - Introduced a concise reporting function to provide a clear summary of the dataset's state before and after cleaning, including the number of rows, columns, and missing values per column.

#### Results:
- The dataset was successfully cleaned, with all missing values addressed. The `Cabin` column was removed, `Age` values were imputed, and remaining rows with missing values were dropped, resulting in a dataset ready for further analysis.

#### Next Steps:
- Proceed with more advanced data analysis, such as feature engineering, data visualization, or building predictive models using the cleaned Titanic dataset.

---

### Error Handling and Debugging Exercises

After cleaning the dataset, we explored various intentional errors and their fixes to enhance our understanding of common coding issues and debugging techniques:

1. **Forget to Import a Library**:
   - **Error Created**: Removed the `import pandas as pd` line, which caused a `NameError: name 'pd' is not defined`.
   - **Solution**: Re-added the `import pandas as pd` line to fix the error.

2. **Incorrect Filename in URL**:
   - **Error Created**: Changed the filename in the URL from `titanic.csv` to `titanics.csv`, resulting in a `FileNotFoundError`.
   - **Solution**: Corrected the filename in the URL back to `titanic.csv`.

3. **Typo in URL Domain or Path**:
   - **Error Created**: Introduced typos in different parts of the URL (e.g., domain name, path), resulting in various errors like `URLError` and `HTTPError 404: Not Found`.
   - **Solution**: Corrected the typos in the URL.

4. **Using a Variable Before Assignment**:
   - **Error Created**: Misnamed the DataFrame variable, e.g., `Titanic_Data` instead of `titanic_data`, causing a `NameError`.
   - **Solution**: Corrected the variable name to `titanic_data`.

5. **Missing Parenthesis in Function Call**:
   - **Error Created**: Removed a closing parenthesis in a function call like `pd.read_csv(url`, resulting in a `SyntaxError: unexpected EOF while parsing`.
   - **Solution**: Added the missing parenthesis to complete the function call.

6. **Mistyped Function Names**:
   - **Error Created**: Misspelled function names like `group_by` instead of `groupby` and `describle` instead of `describe`, causing `AttributeError`.
   - **Solution**: Corrected the function names to `groupby` and `describe`.

7. **Using a Column Name Not in the Data**:
   - **Error Created**: Used incorrect column names in `groupby` and column selection, e.g., `"Sex"` instead of `"sex"` and `"age"` instead of `"Age"`, resulting in `KeyError`.
   - **Solution**: Corrected the column names to match exactly those in the DataFrame.

8. **Forgetting to Put Column Names in Quotes**:
   - **Error Created**: Omitted quotes around column names, such as `groupby(sex)` instead of `groupby("sex")`, causing a `NameError`.
   - **Solution**: Added quotes around column names to ensure they are treated as strings.

### Conclusion:
This exercise demonstrated how to clean a dataset effectively and handle missing values while providing practical experience in identifying, creating, and fixing common programming errors. The process strengthened understanding of debugging techniques and helped develop best practices for data analysis using pandas.
